## A resource of paired ribosomal RNA-depleted and polyA+ RNA-Seq data from primary CD4 T cell of 40 human individuals

### Outline

##### Fig.1 Flow Chart of the RNA-seq experiment and data analysis
##### Fig.2 Summary of key quality control metrics

​	A. fastQC
​	B. Summary of read-mapping
​	C. Reads distribution along gene body

##### Fig.3 Comparison of identified genes and transcripts

​	A. Identified gene number
​	B. Identified gene biotype
​	C. Identified transcript number
​	D. Identified transcript biotype
​	E. Correlation of different method quantification
​	F. GO&KEGG enrichment of method-specific genes

##### Fig.4 Comparison of identified splice junctions

​	A. Splice site motif frequency
​	B. Comparison of novel SJ frequency and fraction
​	C. Novel SJ conservation

##### Fig.5 A multidimensional scaling plot of two method 

​	A. MDS plot
​	B. Dendrogram tree
​	C. MDS plot of protein-coding gene only
​	D. Dendrogram tree of protein-coding gene only

##### Fig.6 eQTL analysis

​	A. QQ plot
​	B. Venn diagram
​	C. Box plot

##### Tab.1 Characteristics of the data

#### Refrence

​	[A comparison between ribo-minus RNA-sequencing and polyA-selected RNA-sequencing](https://www.sciencedirect.com/science/article/pii/S0888754310001746?via%3Dihub)
​	[A resource of ribosomal RNA-depleted RNA-Seq data from different normal adult and fetal human tissues](https://www.nature.com/articles/sdata201563)
​	[Transcriptome sequencing, molecular markers, and transcription factor discovery of Platanus acerifolia in the presence of Corythucha ciliata](https://www.nature.com/articles/sdata2018131)