## Paired rRNA-depleted and polyA-selected RNA-seq from primary T cell of 40 human samples with multi-omics data

## Abstract

â€‹    Both Poly(A) enrichment and ribosomal RNA depletion are commonly used for RNA sequencing. Either has its advantages and disadvantages that may lead to biases in the downstream analyses. To better access these effects, we carried out both ribosomal RNA-depleted and PolyA-selected RNA-Seq for CD4+ T naive cells isolated from 40 healthy individuals from the Blueprint Project. For these 40 individuals, the genomic and epigenetic data were also available. This dataset offers a unique opportunity to understand how library construction influences differential gene expression, alternative splicing and molecular QTL (quantitative loci) analyses for human primary cells.



**Figure 1.** Study design of paired Poly(A) enrichment and ribosomal RNA deletion RNA-sequencing.

**Figure 2.** Summary of key quality control metrics. 

**Figure 3.** Comparison of gene expression identification between PolyA-selected and rRNA-depletion. 

**Figure 4.** Unsupervised clustering analysis. 