
---
title: "Fig.5 A multidimensional scaling plot of two method"
author: "Tang Chao"
date: "9/28/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Content
[A. MDS plot of gene](#A)    
[B. MDS plot of protein-coding gene only](#B)    
[C. MDS plot of isoform](#C)    
[D. MDS plot of protein-coding isoform only](#D)    


## preparation

```{r message=FALSE}
suppressPackageStartupMessages(library(data.table))
gtf <- rtracklayer::readGFF("/mnt/raid61/Personal_data/tangchao/Document/gencode/human/GRCh37/gencode.v30lift37.annotation.gtf")
gtf <- rtracklayer::readGFF("/mnt/raid61/Personal_data/tangchao/Document/gencode/human/GRCh37/gencode.v30lift37.annotation.gtf")
setDT(gtf)
gtf <- gtf[type %in% c("gene", "transcript"), .(seqid, type, start, end, strand, gene_id, gene_type, gene_name, transcript_id, transcript_type)]
gtf[, gene_id:=substr(gene_id, 1, 15)]
gtf[, transcript_id:=substr(transcript_id, 1, 15)]

Gene_biotype <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/Annotation/Gene_biotype.csv", header = F, select = 1:2)
setkey(Gene_biotype, V1)

gtf_gene <- gtf[type == "gene", ]
gtf_gene$gene_biotype <- Gene_biotype[gtf_gene$gene_type, V2]
setkey(gtf_gene, gene_id)

gtf_tx <- gtf[type == "transcript", ]
gtf_tx$transcript_biotype <- Gene_biotype[gtf_tx$transcript_type, V2]
setkey(gtf_tx, transcript_id)
```

## A. MDS plot of gene{#A}

```{r message=FALSE}
library(data.table)

col.p <- "#00AFBB"
col.t <- "#E7B800"

SampInfo_PolyA <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/SampleInfo/PolyA_RNA_sampleInfo.txt")
SampInfo_Total <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/SampleInfo/Total_RNA_sampleInfo.txt")

TPM_PolyA <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/Phenotype/gene/PolyA_RNA_TPM.txt.gz")
TPM_Total <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/Phenotype/gene/Total_RNA_TPM.txt.gz")

TPM_PolyA <- data.frame(TPM_PolyA[, -1], row.names = TPM_PolyA[[1]])
TPM_Total <- data.frame(TPM_Total[, -1], row.names = TPM_Total[[1]])
```

```{r}
# stopifnot(identical(colnames(TPM_PolyA), SampInfo_PolyA$donor_id))
# colnames(TPM_PolyA) <-  SampInfo_PolyA$ID
# 
# stopifnot(identical(colnames(TPM_Total), SampInfo_Total$donor_id))
# colnames(TPM_Total) <-  SampInfo_Total$ID

colnames(TPM_PolyA) <- paste("PolyA", colnames(TPM_PolyA), sep = "_")
colnames(TPM_Total) <- paste("Total", colnames(TPM_Total), sep = "_")

Gene_TPM <- merge(TPM_PolyA, TPM_Total, by = 0)
Gene_TPM <- data.frame(Gene_TPM[, -1], row.names = Gene_TPM[[1]])
Gene_TPM <- log1p(Gene_TPM)
```

```{r fig.width=12}
library(FactoMineR)
library(factoextra)
fviz_dend(hcut(t(Gene_TPM), k = 2, stand = TRUE), rect = TRUE, cex = 1, k_colors = c("#2E9FDF", "#E7B800"))
```

```{r}
library(ggbeeswarm)
library(ggthemes)
res.pca <- PCA(t(Gene_TPM), ncp = 10000, graph = F)
```

```{r}
fviz_screeplot(res.pca, addlabels = TRUE) + 
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14))
```

```{r}
fviz_pca_ind(res.pca, 
             # geom = c("point", "text"), 
             geom = "point", 
             col.ind = "cos2", 
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE # Avoid text overlapping (slow if many points)
             )
```

```{r}
pca_raw_result <- data.frame(res.pca$ind$coord, 
                             ID = colnames(Gene_TPM), 
                             Library = mapply(function(x) x[1], strsplit(colnames(Gene_TPM), "_")))

ggplot(pca_raw_result, aes(x = Dim.1, y = Dim.2, color = Library))+
  geom_point(size=2)+ #Size and alpha just for fun
  scale_color_manual(values = c(col.p, col.t))+ #your colors here
  guides(color = FALSE)+
  xlab(paste("Dim.1(",round(res.pca$eig[,2][1],2),"%)",sep = ""))+
  ylab(paste("Dim.2(",round(res.pca$eig[,2][2],2),"%)",sep = ""))+
  labs(title = "Individuals PCA") +
  theme(panel.background = element_blank(), 
        axis.line = element_blank(), 
        panel.grid = element_line(colour = "grey90"), 
        axis.title = element_text(size = 16), 
        axis.text = element_text(size = 12), legend.position = "none")
```

## B. MDS plot of protein-coding gene only{#B}

```{r}
gtf_gene <- unique(gtf_gene[, .(gene_id, gene_type)])
Index <- gtf_gene[substr(row.names(Gene_TPM), 1, 15), ]$gene_type == "protein_coding"
Gene_TPM <- Gene_TPM[which(Index), ]
```

```{r}
fviz_dend(hcut(t(Gene_TPM), k = 2, stand = TRUE), rect = TRUE, cex = 1, k_colors = c("#2E9FDF", "#E7B800"))
```

```{r}
res.pca <- PCA(t(Gene_TPM), ncp = 1000, graph = F)
fviz_screeplot(res.pca, addlabels = TRUE) + 
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14))
```

```{r}
fviz_pca_ind(res.pca, 
             # geom = c("point", "text"), 
             geom = "point", 
             col.ind = "cos2", 
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE # Avoid text overlapping (slow if many points)
             )
```

```{r}
pca_raw_result <- data.frame(res.pca$ind$coord, 
                             ID = colnames(Gene_TPM), 
                             Library = mapply(function(x) x[1], strsplit(colnames(Gene_TPM), "_")))

ggplot(pca_raw_result, aes(x = Dim.1, y = Dim.2, color = Library))+
  geom_point(size=2)+ #Size and alpha just for fun
  scale_color_manual(values = c(col.p, col.t))+ #your colors here
  guides(color = FALSE)+
  xlab(paste("Dim.1(",round(res.pca$eig[,2][1],2),"%)",sep = ""))+
  ylab(paste("Dim.2(",round(res.pca$eig[,2][2],2),"%)",sep = ""))+
  labs(title = "Individuals PCA") +
  theme(panel.background = element_blank(), 
        axis.line = element_blank(), 
        panel.grid = element_line(colour = "grey90"), 
        axis.title = element_text(size = 16), 
        axis.text = element_text(size = 12), legend.position = "none")
```

## C. MDS plot of isoform{#C}

```{r message=FALSE}
TPM_PolyA <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/Phenotype/transcript/PolyA_RNA_TPM.txt.gz")
TPM_Total <- fread("/mnt/raid61/Personal_data/tangchao/ScientificData/data/Phenotype/transcript/Total_RNA_TPM.txt.gz")

TPM_PolyA <- data.frame(TPM_PolyA[, -1], row.names = TPM_PolyA[[1]])
TPM_Total <- data.frame(TPM_Total[, -1], row.names = TPM_Total[[1]])
```

```{r}
colnames(TPM_PolyA) <- paste("PolyA", colnames(TPM_PolyA), sep = "_")
colnames(TPM_Total) <- paste("Total", colnames(TPM_Total), sep = "_")

Gene_TPM <- merge(TPM_PolyA, TPM_Total, by = 0)
Gene_TPM <- data.frame(Gene_TPM[, -1], row.names = Gene_TPM[[1]])
Gene_TPM <- log1p(Gene_TPM)
```

```{r fig.width=12}
fviz_dend(hcut(t(Gene_TPM), k = 2, stand = TRUE), rect = TRUE, cex = 1, k_colors = c("#2E9FDF", "#E7B800"))
```

```{r}
library(ggbeeswarm)
library(ggthemes)
res.pca <- PCA(t(Gene_TPM), ncp = 10000, graph = F)
```

```{r}
fviz_screeplot(res.pca, addlabels = TRUE) + 
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14))
```

```{r}
fviz_pca_ind(res.pca, 
             # geom = c("point", "text"), 
             geom = "point", 
             col.ind = "cos2", 
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE # Avoid text overlapping (slow if many points)
             )
```

```{r}
pca_raw_result <- data.frame(res.pca$ind$coord, 
                             ID = colnames(Gene_TPM), 
                             Library = mapply(function(x) x[1], strsplit(colnames(Gene_TPM), "_")))

ggplot(pca_raw_result, aes(x = Dim.1, y = Dim.2, color = Library))+
  geom_point(size=2)+ #Size and alpha just for fun
  scale_color_manual(values = c(col.p, col.t))+ #your colors here
  guides(color = FALSE)+
  xlab(paste("Dim.1(",round(res.pca$eig[,2][1],2),"%)",sep = ""))+
  ylab(paste("Dim.2(",round(res.pca$eig[,2][2],2),"%)",sep = ""))+
  labs(title = "Individuals PCA") +
  theme(panel.background = element_blank(), 
        axis.line = element_blank(), 
        panel.grid = element_line(colour = "grey90"), 
        axis.title = element_text(size = 16), 
        axis.text = element_text(size = 12), legend.position = "none")
```

## D. MDS plot of protein-coding isoform only{#D}

```{r}
gtf_tx <- unique(gtf_tx[, .(transcript_id, transcript_type)])
Index <- gtf_tx[substr(row.names(Gene_TPM), 1, 15), ]$transcript_type == "protein_coding"
Gene_TPM <- Gene_TPM[which(Index), ]
```

```{r}
fviz_dend(hcut(t(Gene_TPM), k = 2, stand = TRUE), rect = TRUE, cex = 1, k_colors = c("#2E9FDF", "#E7B800"))
```

```{r}
res.pca <- PCA(t(Gene_TPM), ncp = 1000, graph = F)
fviz_screeplot(res.pca, addlabels = TRUE) + 
  theme(axis.title = element_text(size = 16), axis.text = element_text(size = 14))
```

```{r}
fviz_pca_ind(res.pca, 
             # geom = c("point", "text"), 
             geom = "point", 
             col.ind = "cos2", 
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE # Avoid text overlapping (slow if many points)
             )
```

```{r}
pca_raw_result <- data.frame(res.pca$ind$coord, 
                             ID = colnames(Gene_TPM), 
                             Library = mapply(function(x) x[1], strsplit(colnames(Gene_TPM), "_")))

ggplot(pca_raw_result, aes(x = Dim.1, y = Dim.2, color = Library))+
  geom_point(size=2)+ #Size and alpha just for fun
  scale_color_manual(values = c(col.p, col.t))+ #your colors here
  guides(color = FALSE)+
  xlab(paste("Dim.1(",round(res.pca$eig[,2][1],2),"%)",sep = ""))+
  ylab(paste("Dim.2(",round(res.pca$eig[,2][2],2),"%)",sep = ""))+
  labs(title = "Individuals PCA") +
  theme(panel.background = element_blank(), 
        axis.line = element_blank(), 
        panel.grid = element_line(colour = "grey90"), 
        axis.title = element_text(size = 16), 
        axis.text = element_text(size = 12), legend.position = "none")
```
